File: rb101/rb101-rb109-small_problems/easy6/cute_angles.rb
DEGREE = "\xC2\xB0".freeze
MINUTES_PER_DEGREE = 60
SECONDS_PER_MINUTE = 60
SECONDS_PER_DEGREE = MINUTES_PER_DEGREE * SECONDS_PER_MINUTE

def dms(angle)
  total_seconds = (angle * SECONDS_PER_DEGREE).round
  degrees, remainder = total_seconds.divmod(SECONDS_PER_DEGREE)
  minutes, seconds = remainder.divmod(SECONDS_PER_MINUTE)

  degrees %= 360

  format(%(%d#{DEGREE}%02d'%02d"), degrees, minutes, seconds)
end

p dms(30) == %(30°00'00")
p dms(76.73) == %(76°43'48")
p dms(254.6) == %(254°36'00")
p dms(93.034773) == %(93°02'05")
p dms(0) == %(0°00'00")
p dms(360) == %(360°00'00") || dms(360) == %(0°00'00")

p dms(400) == %(40°00'00")
p dms(-40) == %(320°00'00")
p dms(-420) == %(300°00'00")
p dms(-76.73) == %(283°16'12")

# ALTERNATIVE SOLUTION:

# def dms(angle)
#   degrees = angle
#   minutes = ((angle % 1) * 60)
#   seconds = ((minutes % 1) * 60)

#   degrees %= 360

#   format(%(%d#{DEGREE}%02d'%02d"), degrees.to_i, minutes.to_i, seconds.to_i)
# end

# With this solution some of the calculations may be off by one second.

File: js210/js210-small_problems/easy4/cuteAngles.js
"use strict";

const DEGREES_SYMBOL = '\xB0';
const MINUTES_PER_DEGREES = 60;
const SECONDS_PER_MINUTES = 60;

function dms(degreesFloat) {
  const degreesInt = Math.floor(degreesFloat);

  const minutesFloat = (degreesFloat - degreesInt) * MINUTES_PER_DEGREES;
  const minutesInt = Math.floor(minutesFloat);

  const secondsInt = Math.floor((minutesFloat - minutesInt) * SECONDS_PER_MINUTES);

  return `${degreesInt}${DEGREES_SYMBOL}${padZeroes(minutesInt)}'${padZeroes(secondsInt)}"`;
}

function padZeroes(number) {
  let numString = String(number);
  return numString.padStart(2, '0');
}

console.log(dms(30));           // 30°00'00"
console.log(dms(76.73));        // 76°43'48"
console.log(dms(254.6));        // 254°35'59"
console.log(dms(93.034773));    // 93°02'05"
console.log(dms(0));            // 0°00'00"
console.log(dms(360));          // 360°00'00" or 0°00'00"


// ALTERNATIVE SOLUTION:

// const DEGREE = '\xB0';
// const MINUTES_PER_DEGREE = 60;
// const SECONDS_PER_MINUTE = 60;
// const SECONDS_PER_DEGREE = MINUTES_PER_DEGREE * SECONDS_PER_MINUTE;

// function dms(degreesFloat) {
//   const degreesInt = Math.floor(degreesFloat);
//   const minutes = Math.floor((degreesFloat - degreesInt) * MINUTES_PER_DEGREE);
//   const seconds = Math.floor(
//     (degreesFloat - degreesInt - (minutes / MINUTES_PER_DEGREE)) *
//     SECONDS_PER_DEGREE
//   );

//   return `${String(degreesInt) + DEGREE + padZeroes(minutes)}'${padZeroes(seconds)}"`;
// }

// function padZeroes(number) {
//   const numString = String(number);
//   return numString.length < 2 ? (`0${numString}`) : numString;
// }


// FURTHER EXPLORATION:

// function dms(degreesFloat) {
//   degreesFloat %= 360;

//   if (degreesFloat < 0) {
//     degreesFloat = 360 + degreesFloat;
//   }

//   const degreesInt = Math.floor(degreesFloat);

//   const minutesFloat = (degreesFloat - degreesInt) * MINUTES_PER_DEGREES;
//   const minutesInt = Math.floor(minutesFloat);

//   const secondsInt = Math.floor((minutesFloat - minutesInt) * SECONDS_PER_MINUTES);

//   return `${degreesInt}${DEGREES_SYMBOL}${padZeroes(minutesInt)}'${padZeroes(secondsInt)}"`;
// }

// function padZeroes(number) {
//   let numString = String(number);
//   return numString.padStart(2, '0');
// }

// console.log(dms(-1));     // 359°00'00"
// console.log(dms(400));    // 40°00'00"
// console.log(dms(-40));    // 320°00'00"
// console.log(dms(-420));   // 300°00'00"
// console.log(dms(-76.73)); // 283°16'11"
// console.log(dms(283.27)); // 283°16'11"
